client/grpc/tmservice/service_test.go

```diff
@@ -85,6 +86,39 @@ func (s IntegrationTestSuite) TestQueryLatestBlock() {
 	s.Require().NoError(val.ClientCtx.Codec.UnmarshalJSON(restRes, &blockInfoRes))
 }
 
+func (s IntegrationTestSuite) TestQueryBlockByHash() {
+	val := s.network.Validators[0]
+	node, _ := val.ClientCtx.GetNode()
+	blk, _ := node.Block(context.Background(), nil)
+	blkhash := blk.BlockID.Hash
+
+	tcs := []struct {
+		hash  bytes.HexBytes
+		isErr bool
+		err   string
+	}{
+		{blkhash, false, ""},
+		{bytes.HexBytes("wrong hash"), true, "The length of blcok hash must be 32: invalid request"},
+		{bytes.HexBytes(""), true, "block hash cannot be empty"},
+	}
+
+	for _, tc := range tcs {
+		_, err := s.queryClient.GetBlockByHash(context.Background(), &tmservice.GetBlockByHashRequest{Hash: tc.hash})
+		if tc.isErr {
+			s.Require().Error(err)
+			s.Require().Contains(err.Error(), tc.err)
+		} else {
+			s.Require().NoError(err)
+		}
+	}
+
+	// TODO: Unlike gRPC, REST requests often fail. Need to understand the cause later
```

Comment:
Would you leave the getting failed detail?

Reply:
Is this the same?
```
--- FAIL: TestIntegrationTestSuite (16.96s)
    service_test.go:31: setting up integration test suite
    network.go:174: acquiring test network lock
    network.go:179: created temporary directory: /tmp/TestIntegrationTestSuite2939229296/001/chain-eUzfDc3018637995
    network.go:188: preparing test network...
    network.go:382: starting test network...
    network.go:387: started test network
    --- FAIL: TestIntegrationTestSuite/TestLatestValidatorSet_GRPC (0.00s)
        --- FAIL: TestIntegrationTestSuite/TestLatestValidatorSet_GRPC/no_pagination (0.00s)
            service_test.go:199: 
                	Error Trace:	/home/runner/work/lbm-sdk/lbm-sdk/client/grpc/tmservice/service_test.go:199
                	            				/home/runner/work/lbm-sdk/lbm-sdk/client/grpc/tmservice/suite.go:91
                	Error:      	Received unexpected error:
                	            	rpc error: code = InvalidArgument desc = could not find voter params for height #16: invalid request
                	Test:       	TestIntegrationTestSuite/TestLatestValidatorSet_GRPC/no_pagination
        --- FAIL: TestIntegrationTestSuite/TestLatestValidatorSet_GRPC/with_pagination (0.00s)
            service_test.go:199: 
                	Error Trace:	/home/runner/work/lbm-sdk/lbm-sdk/client/grpc/tmservice/service_test.go:199
                	            				/home/runner/work/lbm-sdk/lbm-sdk/client/grpc/tmservice/suite.go:91
                	Error:      	Received unexpected error:
                	            	rpc error: code = InvalidArgument desc = could not find voter params for height #16: invalid request
                	Test:       	TestIntegrationTestSuite/TestLatestValidatorSet_GRPC/with_pagination
    --- FAIL: TestIntegrationTestSuite/TestLatestValidatorSet_GRPCGateway (0.02s)
        --- FAIL: TestIntegrationTestSuite/TestLatestValidatorSet_GRPCGateway/no_pagination (0.00s)
            service_test.go:232: 
                	Error Trace:	/home/runner/work/lbm-sdk/lbm-sdk/client/grpc/tmservice/service_test.go:232
                	            				/home/runner/work/lbm-sdk/lbm-sdk/client/grpc/tmservice/suite.go:91
                	Error:      	Received unexpected error:
                	            	unknown field "code" in tmservice.GetLatestValidatorSetResponse
                	Test:       	TestIntegrationTestSuite/TestLatestValidatorSet_GRPCGateway/no_pagination
        --- FAIL: TestIntegrationTestSuite/TestLatestValidatorSet_GRPCGateway/with_pagination (0.01s)
            service_test.go:232: 
                	Error Trace:	/home/runner/work/lbm-sdk/lbm-sdk/client/grpc/tmservice/service_test.go:232
                	            				/home/runner/work/lbm-sdk/lbm-sdk/client/grpc/tmservice/suite.go:91
                	Error:      	Received unexpected error:
                	            	unknown field "code" in tmservice.GetLatestValidatorSetResponse
                	Test:       	TestIntegrationTestSuite/TestLatestValidatorSet_GRPCGateway/with_pagination
    service_test.go:48: tearing down integration test suite
    network.go:512: cleaning up test network...
    network.go:535: finished cleaning up test network
    network.go:509: released test network lock
FAIL
FAIL	github.com/line/lbm-sdk/client/grpc/tmservice	17.755s
```

Reply:
yeah it's same issue. even other grpc-gw endpoint raise this issue intermittently

Reply:
Then there are other possible errors as well, so I don't think it's necessary to comment them out.

Reply:
Thank you for your comment. Then, I should probably look at the grpc-gw side later. I was just worried that this intermittent error would cause a situation where PR could not be merged. I will put this problem on the [issue](https://github.com/line/lbm-sdk/issues/703) and reactivate the REST test on this PR.

Reply:
I've confirmed that attached the REST endpoint with Bad Request.
* https://github.com/line/lbm-sdk/blob/main/types/rest/rest.go#L414

```
res, err := http.Get(url) // nolint:gosec
fmt.Println("GetRequest:res.Status:" + res.Status) // Added
```
```
GetRequest:res.Status:400 Bad Request
```

I think it happened `RPC HTTP server stopped` before testing.

Reply:
And I've confirmed that REST was removed from the feature.
* https://github.com/cosmos/cosmos-sdk/pull/9594/files

Then, you can ignore my comments...

Reply:
I've confirmed that. Um, I don't know. But I think it's a timing issue.

```
diff --git a/client/grpc/tmservice/service_test.go b/client/grpc/tmservice/service_test.go
index 2f40d2a8a..e8e82dcf1 100644
--- a/client/grpc/tmservice/service_test.go
+++ b/client/grpc/tmservice/service_test.go
@@ -32,6 +33,7 @@ func (s *IntegrationTestSuite) SetupSuite() {
 
        cfg := network.DefaultConfig()
        cfg.NumValidators = 1
+       cfg.EnableLogging = true
 
        s.cfg = cfg
        s.network = network.New(s.T(), cfg)

diff --git a/types/rest/rest.go b/types/rest/rest.go
index e25e69f71..94d0944b3 100644
--- a/types/rest/rest.go
+++ b/types/rest/rest.go
@@ -12,6 +12,7 @@ import (
        "net/url"
        "strconv"
        "strings"
+       "time"
 
        "github.com/line/ostracon/types"
 
@@ -412,6 +413,7 @@ func ParseQueryParamBool(r *http.Request, param string) bool {
 // An error is returned if the request or reading the body fails.
 func GetRequest(url string) ([]byte, error) {
        res, err := http.Get(url) // nolint:gosec
+       fmt.Println(fmt.Sprintf("GetRequest:Status:[%s]:%s", time.Now(), res.Status))
        if err != nil {
                return nil, err
        }
```
```
I[2022-10-11|20:08:11.159] block generated                              module=state height=15 num_txs=0 generation_time=0.038 tps=0
I[2022-10-11|20:08:11.178] indexed block                                module=txindex height=15
=== CONT  TestIntegrationTestSuite
    service_test.go:49: tearing down integration test suite
    network.go:512: cleaning up test network...
I[2022-10-11|20:08:11.218] Stopping Node service                        module=node0 impl=Node
I[2022-10-11|20:08:11.218] Stopping Node                                module=node0 
I[2022-10-11|20:08:11.218] Stopping EventBus service                    module=events impl=EventBus
I[2022-10-11|20:08:11.218] Stopping PubSub service                      module=pubsub impl=PubSub
I[2022-10-11|20:08:11.218] Stopping IndexerService service              module=txindex impl=IndexerService
I[2022-10-11|20:08:11.218] Stopping P2P Switch service                  module=p2p impl="P2P Switch"
I[2022-10-11|20:08:11.218] Stopping Evidence service                    module=evidence impl=Evidence
I[2022-10-11|20:08:11.218] Stopping StateSync service                   module=statesync impl=StateSync
I[2022-10-11|20:08:11.218] Stopping PEX service                         module=pex impl=PEX
I[2022-10-11|20:08:11.218] Stopping AddrBook service                    module=p2p book=/var/folders/mv/238z3jp135n0xpdhzw373w3h0000gq/T/TestIntegrationTestSuite1282356146/001/chain-g24zvi1460971146/node0/simd/config/addrbook.json impl=AddrBook
I[2022-10-11|20:08:11.218] Stopping Mempool service                     module=mempool impl=Mempool
I[2022-10-11|20:08:11.218] Stopping BlockchainReactor service           module=blockchain impl=BlockchainReactor
I[2022-10-11|20:08:11.218] Stopping Consensus service                   module=consensus impl=ConsensusReactor
I[2022-10-11|20:08:11.218] Stopping State service                       module=consensus impl=ConsensusState
I[2022-10-11|20:08:11.218] Stopping TimeoutTicker service               module=consensus impl=TimeoutTicker
I[2022-10-11|20:08:11.218] Stopping baseWAL service                     module=consensus wal=/var/folders/mv/238z3jp135n0xpdhzw373w3h0000gq/T/TestIntegrationTestSuite1282356146/001/chain-g24zvi1460971146/node0/simd/data/cs.wal/wal impl=baseWAL
I[2022-10-11|20:08:11.218] Saving AddrBook to file                      module=p2p book=/var/folders/mv/238z3jp135n0xpdhzw373w3h0000gq/T/TestIntegrationTestSuite1282356146/001/chain-g24zvi1460971146/node0/simd/config/addrbook.json size=0
I[2022-10-11|20:08:11.247] Stopping Group service                       module=consensus wal=/var/folders/mv/238z3jp135n0xpdhzw373w3h0000gq/T/TestIntegrationTestSuite1282356146/001/chain-g24zvi1460971146/node0/simd/data/cs.wal/wal impl=Group
E[2022-10-11|20:08:11.285] Stopped accept routine, as transport is closed module=p2p numPeers=0
I[2022-10-11|20:08:11.285] Closing rpc listener                         module=node0 listener="&{Listener:0x1400000e6f0 sem:0x1400110bb00 closeOnce:{done:0 m:{state:0 sema:0}} done:0x1400110bb60}"
I[2022-10-11|20:08:11.285] RPC HTTP server stopped                      module=rpc-server err="accept tcp [::]:60256: use of closed network connection"
E[2022-10-11|20:08:11.285] Error serving server                         module=node0 err="accept tcp [::]:60256: use of closed network connection"
I[2022-10-11|20:08:11.286] RPC HTTP server stopped                      module=api-server err="accept tcp [::]:60255: use of closed network connection"
    network.go:535: finished cleaning up test network
    network.go:509: released test network lock
--- FAIL: TestIntegrationTestSuite (16.51s)
=== RUN   TestIntegrationTestSuite/TestQueryBlockByHash
GetRequest:Status:[2022-10-11 20:08:11.218117 +0900 JST m=+16.466768876]:501 Not Implemented
    service_test.go:121: 
        	Error Trace:	/XXXXXXXXXXXX/github.com/line/lbm-sdk/client/grpc/tmservice/service_test.go:121
        	Error:      	Received unexpected error:
        	            	unknown field "code" in tmservice.GetBlockByHashResponse
        	Test:       	TestIntegrationTestSuite/TestQueryBlockByHash
    --- FAIL: TestIntegrationTestSuite/TestQueryBlockByHash (0.00s)


FAIL
```

Reply:
Thank you for your advice. However, the issue occurred not only in the unit test but also in the case of uploading the web server by executing the actual LBM binary, so I think it probably doesn't make sense.

Reply:
The logic itself is a simple correction, so I hope to merge it first and track it through #703 @tnasu

Reply:
Got it! Thanks